store/store.go

```diff
@@ -1391,40 +1413,48 @@ var download = func(ctx context.Context, name, sha3_384, downloadURL string, use
 			continue
 		}
 
-		break
-	}
-	defer resp.Body.Close()
+		defer resp.Body.Close()
 
-	switch resp.StatusCode {
-	case http.StatusOK, http.StatusPartialContent:
-		break
-	case http.StatusUnauthorized:
-		return fmt.Errorf("cannot download non-free snap without purchase")
-	default:
-		return &ErrDownload{Code: resp.StatusCode, URL: resp.Request.URL}
-	}
+		switch resp.StatusCode {
+		case http.StatusOK, http.StatusPartialContent:
+			break
+		case http.StatusUnauthorized:
+			return fmt.Errorf("cannot download non-free snap without purchase")
+		default:
+			return &ErrDownload{Code: resp.StatusCode, URL: resp.Request.URL}
+		}
 
-	if pbar == nil {
-		pbar = &progress.NullProgress{}
-	}
-	pbar.Start(name, float64(resp.ContentLength))
-	mw := io.MultiWriter(w, h, pbar)
-	_, err = io.Copy(mw, resp.Body)
-	pbar.Finished()
-	if err != nil {
-		return err
-	}
+		if pbar == nil {
+			pbar = &progress.NullProgress{}
+		}
+		pbar.Start(name, float64(resp.ContentLength))
+		mw := io.MultiWriter(w, h, pbar)
+		_, err = io.Copy(mw, resp.Body)
+		pbar.Finished()
+		if err != nil {
+			if shouldRetryError(attempt, err) {
+				// error while downloading should resume
+				var seekerr error
+				resume, seekerr = w.Seek(0, os.SEEK_END)
+				if seekerr == nil {
+					continue
+				}
+				// if seek failed, then don't retry end return the original error
+			}
+			return err
+		}
 
-	if cancelled(ctx) {
-		return fmt.Errorf("The download has been cancelled: %s", ctx.Err())
-	}
+		if cancelled(ctx) {
+			return fmt.Errorf("The download has been cancelled: %s", ctx.Err())
+		}
 
-	actualSha3 := fmt.Sprintf("%x", h.Sum(nil))
-	if sha3_384 != "" && sha3_384 != actualSha3 {
-		return fmt.Errorf("sha3-384 mismatch downloading %s: got %s but expected %s", name, actualSha3, sha3_384)
+		actualSha3 := fmt.Sprintf("%x", h.Sum(nil))
+		if sha3_384 != "" && sha3_384 != actualSha3 {
+			return ShaError{name, actualSha3, sha3_384}
+		}
+		return nil
 	}
-
-	return nil
+	return err
```

Comment:
This has that old issue of being error prone. If we add an additional `err := ...` above by mistake, the real error is masked and we return nil here when the strategy says it's time to stop. Let's please make sure this is impossible by either never breaking out of the for loop, or by renaming that error variable so that normal code won't mask it by mistake (`finalErr` or something)